alexa/request/certificate.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+/**
+ * @file Certificate.php
+ * Validate the request signature
+ * Based on code from alexa-app: https://github.com/develpr/alexa-app by Kevin Mitchell
+ * */
+
+namespace Alexa\Request;
+
+use RuntimeException;
+use InvalidArgumentException;
+use DateTime;
+
+class Certificate {
+	const TIMESTAMP_VALID_TOLERANCE_SECONDS = 60;
+	const SIGNATURE_VALID_PROTOCOL = 'https';
+	const SIGNATURE_VALID_HOSTNAME = 's3.amazonaws.com';
+	const SIGNATURE_VALID_PATH = '/echo.api/';
+	const SIGNATURE_VALID_PORT = 443;
+	const ECHO_SERVICE_DOMAIN = 'echo-api.amazon.com';
+	const ENCRYPT_METHOD = 'sha1WithRSAEncryption';
+
+	public $request_id;
+	public $timestamp;
+	/** @var Session */
+	public $session;
+	public $certificate_url;
+	public $certificate_content;
+	public $request_signature;
+	public $request_data;
+	public $app_id;
+
+	/**
+	 * @param type $certificateUri
+	 */
+	public function __construct( $certificate_url, $signature, $app_id = '' ) {
+		$this->certificate_url = $certificate_url;
+		$this->request_signature = $signature;
+		$this->app_id = $app_id;
+	}
+
+	public function validate_request( $request_data ) {
+
+		// Set required http status code 400 for certificate error
+		status_header( 400 );
+
+		$request_parsed = json_decode( $request_data, true );
+		// Validate the entire request by:
+
+		// 1. Checking the timestamp.
+		$this->validate_timestamp( $request_parsed['request']['timestamp'] );
+
+		// 2. Checking if the certificate URL is correct.
+		$this->verify_signature_certificate_url();
+
+		// 3. Checking if the certificate is not expired and has the right SAN
+		$this->validate_certificate();
+
+		// 4. Verifying the request signature
+		$this->validaterequest_signature( $request_data );
+
+		// Set http status code back to 200
+		status_header( 200 );
+	}
+
+	/**
+	 * Check if request is whithin the allowed time.
+	 * @throws InvalidArgumentException
+	 */
+	public function validate_timestamp( $timestamp ) {
+		$now = new DateTime;
+		$timestamp = new DateTime( $timestamp );
+		$difference_in_seconds = $now->getTimestamp() - $timestamp->getTimestamp();
+
+		if ( $difference_in_seconds > self::TIMESTAMP_VALID_TOLERANCE_SECONDS ) {
+			throw new InvalidArgumentException( 'Request timestamp was too old. Possible replay attack.' );
```

Comment:
I'm not going to keep commenting, but since this is a plugin I'd definitely want to see translation functions and a text domain set for the static text throughout.